traffic_ops/testing/api/v4/cachegroups_test.go

```diff
@@ -32,99 +35,393 @@ import (
+						rfc.IfUnmodifiedSince: {time.Now().UTC().Add(-5 * time.Second).Format(time.RFC1123)},
+					},
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusPreconditionFailed)),
+				},
+				"PRECONDITION FAILED when updating with IFMATCH ETAG Header": {
+					GetCacheGroupId(t, "parentCachegroup2"), TOSession, nil,
+					map[string]interface{}{
+						"latitude":  0,
+						"longitude": 0,
+						"name":      "parentCachegroup2",
+						"shortName": "pg2",
+						"typeName":  "MID_LOC",
+						"typeId":    -1,
+					},
+					http.Header{
+						rfc.IfMatch: {rfc.ETag(time.Now().UTC().Add(-5 * time.Second))},
+					},
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusPreconditionFailed)),
+				},
+				"UNAUTHORIZED when NOT LOGGED IN": {
+					GetCacheGroupId(t, "cachegroup1"), NoAuthTOSession, nil, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusUnauthorized)),
+				},
+			},
+			"DELETE": {
+				"NOT FOUND when INVALID ID parameter": {
+					func() int { return 111111 }, TOSession, nil, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusNotFound)),
+				},
+				"UNAUTHORIZED when NOT LOGGED IN": {
+					GetCacheGroupId(t, "cachegroup1"), NoAuthTOSession, nil, nil, nil,
+					utils.CkRequest(utils.HasError(), utils.HasStatus(http.StatusUnauthorized)),
+				},
+			},
+			"GET AFTER CHANGES": {
+				"OK when CHANGES made": {
+					nil, TOSession, nil, nil,
+					http.Header{
+						rfc.IfModifiedSince:   {time.Now().UTC().Add(-5 * time.Second).Format(time.RFC1123)},
+						rfc.IfUnmodifiedSince: {time.Now().UTC().Add(-5 * time.Second).Format(time.RFC1123)},
+					},
+					utils.CkRequest(utils.NoError(), utils.HasStatus(http.StatusOK)),
+				},
+			},
+			"CDNLOCK": {
+				"FORBIDDEN when updating CG when CDN LOCK EXISTS": {
+					nil, nil, nil, nil, nil, nil,
+				},
+			},
+		}
+
+		for method, testCases := range methodTests {
+			t.Run(method, func(t *testing.T) {
+				for name, testCase := range testCases {
+					cg := tc.CacheGroupNullable{}
+					opts := client.NewRequestOptions()
+
+					if testCase.requestParams != nil {
+						for k, v := range testCase.requestParams {
+							if k == "type" {
+								if _, err := strconv.Atoi(v); err != nil {
+									v = strconv.Itoa(GetTypeId(t, v))
+								}
+							}
+							opts.QueryParameters.Add(k, v)
+						}
+					}
+					if testCase.requestHeaders != nil {
+						opts.Header = testCase.requestHeaders
+					}
+					if testCase.requestBody != nil {
+						if _, ok := testCase.requestBody["id"]; ok {
+							testCase.requestBody["id"] = testCase.endpointId()
+						}
+						if typeId, ok := testCase.requestBody["typeId"]; ok {
+							if typeId == -1 {
+								if typeName, ok := testCase.requestBody["typeName"]; ok {
+									testCase.requestBody["typeId"] = GetTypeId(t, typeName.(string))
+								}
+							}
+						}
+						dat, _ := json.Marshal(testCase.requestBody)
```

Comment:
I think ignoring this error could cause debugging confusion when a test uses bad data